README.md
# 💽 osjswowau
**O**pen **S**ource **J**ava**S**cript **WoW** **A**ddon **U**pdater

- ✨ Detects `Interface/AddOns` folder and **installed addons.**
- ⬆️ Checks if **updates are available**. 
- ♻️ **Cross-platform**, supports  **Windows**, **macOs**, **Linux**.
- 📝 Manage with a **configuration file**.
- 🚀 **Parallel** processing of updates.
- 🕵️ Emulates a **real user**, no rate limiting.
- 📦 Builds available on **[Github Releases](https://github.com/jukefr/osjswowau/releases)**.

![demo gif](https://i.imgur.com/I84517Q.gif)

## 📦 Installation
Download from [Github Releases](https://github.com/jukefr/osjswowau/releases) and run it. *(recommended)*

`npm i -g osjswowau` if you prefer to use Nodejs.

## 🗃 Addon Sources
- **CurseForge addons** (latest version, md5 checked)
- **WoWInterface addons** (latest version, no hashes available)
- **Elvui, Tukui and Tukui addons** (latest version, no hashes available)
- **TradeSkillMaster and TradeSkillMaster_Apphelper** (latest version, no hashes available)


## 📝 Configuration
The default settings should work out of the box for most people.

When encountering a timeout error, a second or third attempt at launching osjswowau should work.

If not, you might need to adjust some configuration values.

`__utils.js.schema` contains useful comments to explain the variables.

### ⚡️ CurseForge addons - https://www.curseforge.com/wow/addons
CurseForge addon names go in `config.addons.curse`.

To find the name of an addon from the URL :
```
https://www.curseforge.com/wow/addons/plater-nameplates/...
-> plater-nameplates
```

### ⚡️ WoWInterface addons - https://www.wowinterface.com/
WoWInterface addon names go in `config.addons.wowinterface`

To find the name of an addon from the URL :
```
https://www.wowinterface.com/downloads/info24608-Hekili.html
https://www.wowinterface.com/downloads/download24608-Hekili
-> 24608-Hekili
```

### ⚡️ Elvui, Tukui and Tukui addons - https://www.tukui.org
Tukui addon names go in `config.addons.tukui.addons`

To find the name of an addon from the URL :
```
https://www.tukui.org/addons.php?id=137
-> 137
```
Elvui is enabled by setting `config.addons.tukui.elvui` to `true`.

Tukui is enabled by setting `config.addons.tukui.tukui` to `true`.

### ⚡️ TradeSkillMaster and TradeSkillMaster_Apphelper - https://www.tradeskillmaster.com
TradeSkillMaster and TradeSkillMaster_Apphelper are enabled by setting `config.addons.tsm` to `true`.

**Note** : you need the [corresponding client](https://www.tradeskillmaster.com/install). (for now)

On linux just run it in the same wine container with the same prefix and everything should work just fine.

## 🚑 Debug
The debug mode can help you learn the location of the configuration file.

It also display more information about errors which may be useful for reporting issues.

To enable the debug mode either set `config.debug` to `true`.

Or run the command with the `DEBUG` environment variable : `DEBUG=1 osjswowau`

## ✅ Future Plans
- [ ] classic/retail toggle (currently only retail)
- [ ] wireshark the tsm client to see how it gets the latest values
- [ ] todo implement own bar system that doesnt get fucked when the barArray.length is greater than the console.height....
- [ ] streamify everything
- [ ] remove duplicates from curse/wowinterface now that we have map
- [ ] add "search" now that we have detection with duckduckgo
- [ ] checksum on extract to check versions matchs/addon integrity

CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v6.1.4](https://github.com/jukefr/osjswowau/compare/v6.1.3...v6.1.4)

- just try catch this [`18c310c`](https://github.com/jukefr/osjswowau/commit/18c310ceaac1979d1420f9982778e44a874b2b32)
- just try catch this [`6dba495`](https://github.com/jukefr/osjswowau/commit/6dba49560af154bce96b7c9c516c5dd527460f3a)

#### [v6.1.3](https://github.com/jukefr/osjswowau/compare/v6.1.2...v6.1.3)

> 30 December 2020

- chore: release v6.1.3 [`ed2ee6d`](https://github.com/jukefr/osjswowau/commit/ed2ee6d5193e1db2c84bcee88371c8a07ebc8147)
- another one [`08423f5`](https://github.com/jukefr/osjswowau/commit/08423f55001fc3d18060f5de653f407d855d1ada)

#### [v6.1.2](https://github.com/jukefr/osjswowau/compare/v6.1.1...v6.1.2)

> 30 December 2020

- chore: release v6.1.2 [`2f5770e`](https://github.com/jukefr/osjswowau/commit/2f5770e7e465ece768747340018492cfde71dfd7)
- another one [`e7665dd`](https://github.com/jukefr/osjswowau/commit/e7665dd8f79d2c47a425e695dd5677742dc4d58e)

#### [v6.1.1](https://github.com/jukefr/osjswowau/compare/v6.1.0...v6.1.1)

> 30 December 2020

- chore: release v6.1.1 [`89c40bd`](https://github.com/jukefr/osjswowau/commit/89c40bd90bfe0d01d1d81c5601102a6333759802)
- another one [`7f29888`](https://github.com/jukefr/osjswowau/commit/7f29888d0036e2ed4530a6ef41b732991e8b1198)

#### [v6.1.0](https://github.com/jukefr/osjswowau/compare/v6.0.4...v6.1.0)

> 30 December 2020

- chore: release v6.1.0 [`b0740c6`](https://github.com/jukefr/osjswowau/commit/b0740c6aebc7bcc13980e12fc9cf6e69eacbabeb)
- add fucked windows paths [`d2d0163`](https://github.com/jukefr/osjswowau/commit/d2d016359b3c670f5a3c89dea65f9c24e40ba97f)

#### [v6.0.4](https://github.com/jukefr/osjswowau/compare/v6.0.3...v6.0.4)

> 17 December 2020

- chore: release v6.0.4 [`c2afff0`](https://github.com/jukefr/osjswowau/commit/c2afff086b35d73199d5794322fab5b04ba7f9e7)
- update demo gif [`529a576`](https://github.com/jukefr/osjswowau/commit/529a576828a961126250b6cf2a8be014a6e51702)

#### [v6.0.3](https://github.com/jukefr/osjswowau/compare/v6.0.2...v6.0.3)

> 17 December 2020

- shuffle the Qs [`6530483`](https://github.com/jukefr/osjswowau/commit/65304839cb7852a9d868cdd78b26a65d662bf336)
- chore: release v6.0.3 [`7099287`](https://github.com/jukefr/osjswowau/commit/709928797723df6d915dc7037a70a50a8862f107)

#### [v6.0.2](https://github.com/jukefr/osjswowau/compare/v6.0.1...v6.0.2)

> 17 December 2020

- check toc exists for now [`20d057f`](https://github.com/jukefr/osjswowau/commit/20d057f146bd6a1a7c76680f310aa7efda233114)
- chore: release v6.0.2 [`68f9ae1`](https://github.com/jukefr/osjswowau/commit/68f9ae17405830a92e3fc43c472cd3e78d25e4cd)
- fix tests [`9f87ba0`](https://github.com/jukefr/osjswowau/commit/9f87ba01f0ece29e18b2ab502d79fdff67b56432)

#### [v6.0.1](https://github.com/jukefr/osjswowau/compare/v6.0.0...v6.0.1)

> 17 December 2020

- fallbacks [`34ecc3c`](https://github.com/jukefr/osjswowau/commit/34ecc3cd0ece5fac78a13b8ca98da8881e8770cd)
- chore: release v6.0.1 [`e4eb03a`](https://github.com/jukefr/osjswowau/commit/e4eb03a7c6409afd7ed4e7dc9aa2eb41f2bd2ba3)

### [v6.0.0](https://github.com/jukefr/osjswowau/compare/v5.0.13...v6.0.0)

> 16 December 2020

- massive rework of how toc stuff is handled [`1557872`](https://github.com/jukefr/osjswowau/commit/1557872560380f2d1c41fe0c924b1600d038b64c)
- chore: release v6.0.0 [`189bafc`](https://github.com/jukefr/osjswowau/commit/189bafcd1dd46b7fed01b45e1cd46cd98f9a80ed)

#### [v5.0.13](https://github.com/jukefr/osjswowau/compare/v5.0.12...v5.0.13)

> 16 December 2020

- chore: release v5.0.13 [`0c26eea`](https://github.com/jukefr/osjswowau/commit/0c26eea9236309f7cced430293a6229590bd0007)
- better curse url detection query [`d5c23e7`](https://github.com/jukefr/osjswowau/commit/d5c23e721da39277bd396c359366adcc0021f070)
- update package desc [`79fb5b2`](https://github.com/jukefr/osjswowau/commit/79fb5b295465fd836fcb0ffc00f75fa70df67c8b)

#### [v5.0.12](https://github.com/jukefr/osjswowau/compare/v5.0.11...v5.0.12)

> 15 December 2020

- chore: release v5.0.12 [`58c5fa5`](https://github.com/jukefr/osjswowau/commit/58c5fa5614fc6cb9b6cc3652cb3ac30cd3873872)
- fix typo in docs [`11a3af7`](https://github.com/jukefr/osjswowau/commit/11a3af7e7da88f33ce0feca9c185d0af961af0d6)

#### [v5.0.11](https://github.com/jukefr/osjswowau/compare/v5.0.10...v5.0.11)

> 15 December 2020

- chore: release v5.0.11 [`0edf3f2`](https://github.com/jukefr/osjswowau/commit/0edf3f29c4191f45dc84edcac00ee0baf42e788f)
- update docs [`09f327e`](https://github.com/jukefr/osjswowau/commit/09f327ea3ed52bb69663de8aef7dbe88305bb6e4)

#### [v5.0.10](https://github.com/jukefr/osjswowau/compare/v5.0.9...v5.0.10)

> 15 December 2020

- stop bars [`b0fdafe`](https://github.com/jukefr/osjswowau/commit/b0fdafe4aca349d60e8e7df2264ac9ab0ee728c5)
- chore: release v5.0.10 [`b18bf5a`](https://github.com/jukefr/osjswowau/commit/b18bf5ad49c8da51c2a4114f1d28da16f7ba04d9)
- will fix testing later [`0a8aa09`](https://github.com/jukefr/osjswowau/commit/0a8aa090cc640c4b51452434af35dba5d4514dae)

#### [v5.0.9](https://github.com/jukefr/osjswowau/compare/v5.0.8...v5.0.9)

> 14 December 2020

- chore: release v5.0.9 [`d7d5585`](https://github.com/jukefr/osjswowau/commit/d7d55854ea09e31b6c6102517b2df72fce9f0fff)
- update docs [`4579ae9`](https://github.com/jukefr/osjswowau/commit/4579ae9807c457bfbb17f10793c71b97fc2afcb0)

#### [v5.0.8](https://github.com/jukefr/osjswowau/compare/v5.0.7...v5.0.8)

> 14 December 2020

- major improvement to updates [`928d2da`](https://github.com/jukefr/osjswowau/commit/928d2da8d85860698a09667f40b05df654f0859b)
- curse version checks [`8ad1d87`](https://github.com/jukefr/osjswowau/commit/8ad1d877dabb3b1dbf2acffed1f5451337be0bf1)

_utils.js
const chalk = require("chalk");
const { existsSync, statSync, createReadStream, rmdirSync, unlinkSync } = require("fs");
const glob = require("glob-promise");
const https = require("https");
const md5File = require("md5-file");
const unzipper = require("unzipper");
const pkg = require("./package.json");

const schema = {
  fresh: { type: "boolean", default: true }, // used on first start, should remain disabled after
  headless: { type: "boolean", default: true }, // hide chromium windows
  concurrency: { type: "number", maximum: 10, minimum: 1, default: 5 }, // amount of addons that can be updated at the same time ("threads")
  addonPath: {
    type: "string",
    default: process.platform.includes("win") ? "C:\\path\\to\\addons\\folder\\..." : "path/to/addons/folder",
  },
  waitForKey: { type: "boolean", default: !!process.__nexe }, // wait for a key to continue, enabled by default on nexe
  timeout: {
    // how long an action can take (goto, click, wait, etc) in ms
    type: "number",
    maximum: 300 * 1000,
    minimum: 10 * 1000,
    default: 30 * 1000,
  },
  polling: {
    // how long timeout loops wait after each check in ms
    type: "number",
    maximum: 10 * 1000,
    minimum: 250,
    default: 1000,
  },
  delay: {
    // applied after every navigation action (goto, click, etc) in ms
    type: "number",
    maximum: 10 * 1000,
    minimum: 250,
    default: 2 * 1000,
  },
  addons: {
    type: "object",
    properties: {
      curse: {
        type: "array",
        items: { type: "string", default: "azeroth-auto-pilot" },
        default: [
          "azeroth-auto-pilot",
          "big-wigs",
          "details",
          "little-wigs",
          "pawn",
          "plater-nameplates",
          "weakauras-2",
        ],
      },
      tukui: {
        tukui: { type: "boolean", default: false },
        elvui: { type: "boolean", default: true },
        addons: {
          type: "array",
          items: { type: "number", default: 3 },
          default: [137, 38, 3],
        },
      },
      tsm: { type: "boolean", default: false },
      wowinterface: {
        type: "array",
        items: { type: "string", default: "24608-Hekili" },
        default: [],
      },
    },
    default: {},
  },
};

const deleteFile = (path) => unlinkSync(path);

// TODO: find an automatic way to do this....
const getRevision = (p) => {
  if (p === "linux") return "812859";
  if (p === "mac") return "812851";
  if (p === "win64" || p === "win32") return "812899";
  throw new Error("unsupported OS currently sorry");
};

const getLatestTag = async () =>
  new Promise((resolve, reject) =>
    https
      .get(
        "https://api.github.com/repos/jukefr/osjswowau/tags",
        {
          headers: { "User-Agent": "Mozilla/5.0" },
        },
        (resp) => {
          let data = "";

          resp.on("data", (chunk) => {
            data += chunk;
          });

          resp.on("end", () => resolve(JSON.parse(data)[0]));
        }
      )
      .on("error", (err) => reject(err))
  );

const migrations = {
  "2.2.0": (store) => {
    const curse = store.get("addons.curse");
    if (curse) {
      const removeTSM = curse.filter((i) => {
        const includes = ["tradeskillmaster_apphelper", "tradeskill-master"].includes(i);
        if (!includes) {
          return i;
        }
        store.set("addons.tsm", includes);
        return false;
      });
      return store.set("addons.curse", removeTSM);
    }
    return null;
  },
  "2.2.2": (store) => {
    const addonPath = store.get("realpath");
    if (addonPath) {
      store.delete("realpath");
      store.set("addonPath", addonPath);
    }

    const delay = store.get("waitAfterNavig");
    if (delay) {
      store.delete("waitAfterNavig");
      store.set("delay", delay);
    }
  },
  "3.1.0": (store) => {
    const elvui = store.get("addons.elvui");
    if (Array.isArray(elvui)) {
      store.set("addons.tukui.addons", elvui);
      store.set("addons.tukui.elvui", true);
      store.delete("addons.elvui");
    }
  },
};

const cleanTmps = async (tmp) => {
  const tmps = await glob(`${tmp}-*`);
  const queue = tmps.map((t) => rmdirSync(t, { recursive: true }));
  return Promise.all(queue);
};

class FreshStartError extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "FreshStartError";
  }
}

class WaitTimeoutError extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = "WaitTimeoutError";
  }
}

const firstStart = (config) => {
  if (config.get("fresh")) {
    console.log("First run or configuration updated.");
    console.log(`Please edit ${chalk.yellow(config.path)} to match your needs.`);
    if (process.platform === "win32" || process.platform === "win64")
      console.log(
        `Make sure to use double backslashes ${chalk.yellow("\\\\")} to escape the ${chalk.yellow(
          "addonPath"
        )} (AddOns folder) variable.`
      );
    if (process.platform === "win32" || process.platform === "win64")
      console.log(`ie. ${chalk.yellow('"C:\\\\Program Files\\\\..."')}`);
    config.set("fresh", false);
    throw new FreshStartError();
  }
};

const delay = (d) =>
  new Promise((resolve) => {
    setTimeout(resolve, d);
  });

const errorLogic = (err, debug) => {
  switch (err.constructor) {
    case FreshStartError:
      console.log("");
      console.log(chalk.red("This is your first run. Terminating so you can edit config."));
      if (debug) console.log(chalk.red("cause"), err.cause || err);
      break;
    case SyntaxError:
      console.log("");
      console.log(chalk.red("Your configuration file probably has an incorrect syntax."));
      if (debug) console.log(chalk.red("cause"), err.cause || err);
      break;
    default:
      console.log("");
      console.log(chalk.red("Something went wrong. Usually a re-run of the command should work."));
      console.log(
        chalk.red(`Otherwise enable debug mode to learn more. (start with ${chalk.bold(chalk.red("DEBUG=1"))} env)`)
      );
      console.log(chalk.italic(chalk.bold(chalk.red(err.message))));
      if (debug) console.log(chalk.red(err.constructor.name));
      if (debug) console.log(chalk.red("trace"), err.stack || err);
  }
};

const waitToContinue = () => {
  console.log("");
  console.log(chalk.bold("Press any key to continue..."));
  process.stdin.setRawMode(true);
  return new Promise((resolve) =>
    process.stdin.once("data", () => {
      process.stdin.setRawMode(false);
      return resolve();
    })
  );
};

const endLogic = async (config) => {
  const latest = await getLatestTag();
  if (latest) {
    const latestName = latest.name.replace("v", "");
    console.log(chalk.bold(chalk.green("osjswowau")), `v${chalk.bold(pkg.version)}`, "finished");
    if (`${latestName}` !== `${pkg.version}`) {
      console.log("");
      console.log(
        "new version",
        chalk.bold(chalk.green(latestName)),
        "detected, you are running",
        chalk.bold(chalk.red(pkg.version))
      );
      console.log("please run", chalk.bold(chalk.green('"npm i -g osjswowau"')), "to update");
      console.log(
        "or download the latest binary build from",
        chalk.bold(chalk.green("https://github.com/jukefr/osjswowau/releases"))
      );
    }
  }
  if (process.__nexe) {
    return waitToContinue();
  }
  if (config && config.get("waitForKey")) {
    return waitToContinue();
  }
  return null;
};

const errorLogicWrapper = async (err, config, debug) => {
  errorLogic(err, debug);
  await endLogic(config);
  process.exit(1);
};

const waitMd5 = async (config, m, name, tmp) => {
  const start = Date.now();
  while (Date.now() - start < config.get("timeout")) {
    const [fname] = await glob(`${tmp}-${name}/*.zip`);
    if (existsSync(fname)) {
      const md5 = await md5File(fname);
      if (md5 === m) {
        return fname;
      }
    }
    await delay(config.get("polling"));
  }
  throw new WaitTimeoutError();
};

const waitFile = async (config, f, m, tmp) => {
  const start = Date.now();
  let size;

  while (Date.now() - start < config.get("timeout")) {
    const [fname] = await glob(`${tmp}-${m}/*.zip`);
    if (existsSync(fname)) {
      if (size && size !== 0 && size === statSync(fname).size) return fname;
      size = statSync(fname).size;
    }
    await delay(config.get("polling"));
  }
  throw new WaitTimeoutError();
};

const unzip = (config, filename) =>
  new Promise((resolve, reject) =>
    createReadStream(filename)
      .on("error", (err) => reject(err))
      .pipe(unzipper.Extract({ path: config.get("addonPath") }))
      .on("close", (err) => (err ? reject(err) : resolve()))
      .on("error", (err) => reject(err))
  );

module.exports = {
  getRevision,
  unzip,
  waitFile,
  waitMd5,
  firstStart,
  schema,
  delay,
  cleanTmps,
  migrations,
  getLatestTag,
  deleteFile,
  FreshStartError,
  WaitTimeoutError,
  errorLogic,
  endLogic,
  errorLogicWrapper,
};
